File: apple/taOSc/PairingService.swift
import Foundation

enum PairingService {
    static func createPairRequest(
        baseURL: URL,
        platform: String,
        displayName: String
    ) async throws -> PairRequestResponse {
        let url = baseURL.appendingPathComponent("api/devices/pair-requests")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "platform": platform,
            "display_name": displayName
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw PairingError.unreachable
        }
        return try JSONDecoder().decode(PairRequestResponse.self, from: data)
    }

    static func pollPairRequest(
        baseURL: URL,
        requestId: String
    ) async throws -> PairRequestStatus {
        let url = baseURL.appendingPathComponent("api/devices/pair-requests/\(requestId)")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw PairingError.unreachable
        }

        switch httpResponse.statusCode {
        case 200:
            let pollResponse = try JSONDecoder().decode(PairRequestPollResponse.self, from: data)
            guard let status = pollResponse.requestStatus else {
                throw PairingError.invalidResponse
            }
            return status
        case 404:
            throw PairingError.invalidResponse
        default:
            throw PairingError.unreachable
        }
    }

    static func taOSgoJoin(
        email: String,
        password: String,
        deviceName: String?
    ) async throws -> TaOSgoJoinResponse {
        // tsk-5rukhy: website endpoint not yet shipped; build against documented shape.
        guard let url = URL(string: "https://taos.my/api/taosgo/app-join") else {
            throw PairingError.unreachable
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        var body: [String: Any] = [
            "email": email,
            "password": password
        ]
        if let deviceName = deviceName, !deviceName.isEmpty {
            body["device_name"] = deviceName
        }
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw PairingError.unreachable
        }
        return try JSONDecoder().decode(TaOSgoJoinResponse.self, from: data)
    }
}

File: android/app/src/main/java/com/taosc/taosc/PairingService.kt
package com.taosc.taosc

import java.net.HttpURLConnection
import java.net.URL
import org.json.JSONObject

interface HttpClient {
    fun post(url: String, body: String, headers: Map<String, String> = emptyMap()): HttpResponse
    fun get(url: String, headers: Map<String, String> = emptyMap()): HttpResponse
    fun patch(url: String, body: String, headers: Map<String, String> = emptyMap()): HttpResponse
}

data class HttpResponse(
    val code: Int,
    val body: String
)

    class DefaultHttpClient : HttpClient {
        override fun post(url: String, body: String, headers: Map<String, String>): HttpResponse {
            val connection = URL(url).openConnection() as HttpURLConnection
            connection.requestMethod = "POST"
            headers.forEach { (key, value) -> connection.setRequestProperty(key, value) }
            connection.doOutput = true
            connection.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            
            val responseCode = connection.responseCode
            val responseBody = if (responseCode in 200..299) {
                connection.inputStream.bufferedReader().use { it.readText() }
            } else {
                connection.errorStream?.bufferedReader()?.use { it.readText() } ?: ""
            }
            
            return HttpResponse(responseCode, responseBody)
        }
        
        override fun get(url: String, headers: Map<String, String>): HttpResponse {
            val connection = URL(url).openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            headers.forEach { (key, value) -> connection.setRequestProperty(key, value) }
            
            val responseCode = connection.responseCode
            val responseBody = if (responseCode in 200..299) {
                connection.inputStream.bufferedReader().use { it.readText() }
            } else {
                connection.errorStream?.bufferedReader()?.use { it.readText() } ?: ""
            }
            
            return HttpResponse(responseCode, responseBody)
        }
        
        override fun patch(url: String, body: String, headers: Map<String, String>): HttpResponse {
            val connection = URL(url).openConnection() as HttpURLConnection
            connection.requestMethod = "PATCH"
            headers.forEach { (key, value) -> connection.setRequestProperty(key, value) }
            connection.doOutput = true
            connection.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            
            val responseCode = connection.responseCode
            val responseBody = if (responseCode in 200..299) {
                connection.inputStream.bufferedReader().use { it.readText() }
            } else {
                connection.errorStream?.bufferedReader()?.use { it.readText() } ?: ""
            }
            
            return HttpResponse(responseCode, responseBody)
        }
    }

class PairingService(private val httpClient: HttpClient = DefaultHttpClient()) {
    fun createPairRequest(baseUrl: String, platform: String, displayName: String): PairRequestResponse {
        val url = "$baseUrl/api/devices/pair-requests"
        val body = Json.buildObject(
            "platform" to platform,
            "display_name" to displayName
        )
        
        val response = httpClient.post(url, body, mapOf("Content-Type" to "application/json"))
        
        if (response.code !in 200..299) {
            throw PairingError.Unreachable
        }
        
        val json = try {
            JSONObject(response.body)
        } catch (e: org.json.JSONException) {
            throw PairingError.InvalidResponse
        }
        
        return PairRequestResponse(
            pairRequestId = json.optString("pair_request_id").ifEmpty { null } ?: throw PairingError.InvalidResponse,
            verifyCode = json.optString("verify_code").ifEmpty { null } ?: throw PairingError.InvalidResponse
        )
    }
    
    fun pollPairRequest(baseUrl: String, requestId: String): PairRequestStatus {
        val url = "$baseUrl/api/devices/pair-requests/$requestId"
        val response = httpClient.get(url)
        
        when (response.code) {
            200 -> {
                val json = try {
                    JSONObject(response.body)
                } catch (e: org.json.JSONException) {
                    throw PairingError.InvalidResponse
                }
                val pollResponse = PairRequestPollResponse(
                    id = json.optString("id").ifEmpty { null } ?: throw PairingError.InvalidResponse,
                    status = json.optString("status").ifEmpty { null } ?: throw PairingError.InvalidResponse,
                    scopedToken = if (json.isNull("scoped_token")) null else json.optString("scoped_token").ifEmpty { null }
                )
                return pollResponse.requestStatus ?: throw PairingError.InvalidResponse
            }
            404 -> throw PairingError.InvalidResponse
            else -> throw PairingError.Unreachable
        }
    }
    
    fun updatePushToken(baseUrl: String, deviceId: String, pushToken: String, scopedToken: String) {
        val url = "$baseUrl/api/devices/$deviceId/push-token"
        val body = Json.buildObject(
            "push_token" to pushToken
        )
        
        val response = patch(url, body, mapOf(
            "Content-Type" to "application/json",
            "Authorization" to "Bearer $scopedToken"
        ))
        
        if (response.code !in 200..299) {
            throw PairingError.Unreachable
        }
    }
    
    private fun patch(url: String, body: String, headers: Map<String, String>): HttpResponse {
        // Implement PATCH using HttpURLConnection
        val connection = URL(url).openConnection() as HttpURLConnection
        connection.requestMethod = "PATCH"
        headers.forEach { (key, value) -> connection.setRequestProperty(key, value) }
        connection.doOutput = true
        connection.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
        
        val responseCode = connection.responseCode
        val responseBody = if (responseCode in 200..299) {
            connection.inputStream.bufferedReader().use { it.readText() }
        } else {
            connection.errorStream?.bufferedReader()?.use { it.readText() } ?: ""
        }
        
        return HttpResponse(responseCode, responseBody)
    }
}

private object Json {
    fun buildObject(vararg pairs: Pair<String, String>): String {
        return pairs.joinToString(prefix = "{", postfix = "}") { (key, value) ->
            "\"${escape(key)}\":\"${escape(value)}\""
        }
    }
    
    private fun escape(str: String): String {
        return str.replace("\\", "\\\\").replace("\"", "\\\"")
    }
}
